selfdrive/locationd/locationd.cc

```diff
@@ -0,0 +1,344 @@
+    this->calibrated = log.getCalStatus() == 1;
+  }
+}
+
+void Localizer::reset_kalman(double current_time) {
+  VectorXd init_x = this->kf->get_initial_x();
+  this->reset_kalman(current_time, init_x.segment<4>(3), init_x.head(3));
+}
+
+void Localizer::reset_kalman(double current_time, VectorXd init_orient, VectorXd init_pos) {
+  // too nonlinear to init on completely wrong
+  VectorXd init_x = this->kf->get_initial_x();
+  MatrixXdr init_P = this->kf->get_initial_P();
+  init_x.segment<4>(3) = init_orient;
+  init_x.head(3) = init_pos;
+
+  this->kf->init_state(init_x, init_P, current_time);
+
+  this->gyro_counter = 0;
+  this->acc_counter = 0;
+  this->speed_counter = 0;
+  this->cam_counter = 0;
+}
+
+int Localizer::locationd_thread() {
+  const std::initializer_list<const char *> service_list =
+      { "gpsLocationExternal", "sensorEvents", "cameraOdometry", "liveCalibration", "carState" };
+  SubMaster sm(service_list, nullptr, { "gpsLocationExternal" });
+  PubMaster pm({ "liveLocationKalman" });
+
+  Params params;
+
+  while (!do_exit) {
+    bool updatedCameraOdometry = false;
+    sm.update();
+    for (const char* service : service_list) {
+      if (sm.updated(service) && sm.valid(service)) {
+        cereal::Event::Reader& log = sm[service];
+        double t = log.getLogMonoTime() * 1e-9;
+
+        if (log.isSensorEvents()) {
+          this->handle_sensors(t, log.getSensorEvents());
+        } else if (log.isGpsLocationExternal()) {
+          this->handle_gps(t, log.getGpsLocationExternal());
+        } else if (log.isCarState()) {
+          this->handle_car_state(t, log.getCarState());
+        } else if (log.isCameraOdometry()) {
+          this->handle_cam_odo(t, log.getCameraOdometry());
+          updatedCameraOdometry = true;
+        } else if (log.isLiveCalibration()) {
+          this->handle_live_calib(t, log.getLiveCalibration());
+        } else {
+          std::cout << "invalid event" << std::endl;
+        }
+      }
+    }
+
+    if (updatedCameraOdometry) {
+      uint64_t t = sm["cameraOdometry"].getLogMonoTime();
+
+      MessageBuilder msg_builder;
+      auto evt = msg_builder.initEvent();
+      evt.setLogMonoTime(t);
+      auto liveLoc = evt.initLiveLocationKalman();
+      this->liveLocationMsg(liveLoc);
+      liveLoc.setInputsOK(sm.allAliveAndValid());
+      liveLoc.setSensorsOK(sm.alive("sensorEvents") && sm.valid("sensorEvents"));
+      liveLoc.setGpsOK((t / 1e9) - this->last_gps_fix < 1.0);
+      pm.send("liveLocationKalman", msg_builder);
+
+      if (sm.frame % 1200 == 0 && liveLoc.getGpsOK()) {  // once a minute
+        std::string lastGPSPosJSON = util::string_format("{\"latitude\": %.15f, \"longitude\": %.15f, \"altitude\": %.15f}",
+          liveLoc.getPositionGeodetic().getValue()[0], liveLoc.getPositionGeodetic().getValue()[1], liveLoc.getPositionGeodetic().getValue()[2]);
+        params.put("LastGPSPosition", lastGPSPosJSON);
```

Comment:
We should use an async write here. Filesystem access can block the whole thread for quite a while. Let's put this on the list for the cleanup PR.